tempesta_fw/http_msg.c

```diff
@@ -155,22 +156,50 @@ __hdr_is_singular(const TfwStr *hdr)
  * an HTTP message. Duplicate of a singular header fields is a bug worth
  * blocking the whole HTTP message.
  */
-static int
+static unsigned int
 __hdr_lookup(TfwHttpMsg *hm, const TfwStr *hdr)
 {
-	int id;
+	unsigned int id;
 	TfwHttpHdrTbl *ht = hm->h_tbl;
 
 	for (id = TFW_HTTP_HDR_RAW; id < ht->off; ++id) {
 		TfwStr *h = &ht->tbl[id];
 		/* There is no sense to compare against all duplicates. */
 		if (h->flags & TFW_STR_DUPLICATE)
 			h = TFW_STR_CHUNK(h, 0);
-		if (tfw_stricmpspn(hdr, h, ':'))
-			continue;
+		if (!tfw_stricmpspn(hdr, h, ':'))
+			break;
+	}
+	if (id <  ht->off) {
 		if (__hdr_is_singular(hdr))
 			hm->flags |= TFW_HTTP_FIELD_DUPENTRY;
-		break;
+	}
+
+	return id;
+}
+
+/**
+ * Look up for header named @name, same as @__hdr_lookup, but @name does
+ * not contain ':'.  Checkes for header singularity is skipped, since not needed
+ * outside parser.
+ */
+static unsigned int
+__hdr_lookup_hdr_name(TfwHttpMsg *hm, const TfwStr *name)
+{
+	unsigned int id;
+	TfwHttpHdrTbl *ht = hm->h_tbl;
+
+	for (id = TFW_HTTP_HDR_RAW; id < ht->off; ++id) {
+		TfwStr *h = &ht->tbl[id];
+		/* Skip if header is already marked as hop-by-hop */
+		if (h->flags & TFW_STR_HBH_HDR)
+			continue;
+		/* There is no sense to compare against all duplicates. */
+		if (h->flags & TFW_STR_DUPLICATE)
+			h = TFW_STR_CHUNK(h, 0);
+		if (tfw_str_eq(name, h, name->len)
+		    && (':' == tfw_str_at_index(h, name->len)))
+			break;
```

Comment:
Please see the comment above. It looks to me that there's no need to introduce two new TwfStr{} strings function `tfw_str_eq()` and `tfw_str_at_index()`. At least not until that functionality is really needed.

The core of this function should be made a separate function that can be used by `__hdr_lookup()`. That separate function would return the `id` if it's found, and it can be used directly in place of this `__hdr_lookup_hdr_name()`.